File: test/cli.test.ts
import { execFile } from 'node:child_process';
import { createServer } from 'node:http';
import type { Server } from 'node:http';
import type { AddressInfo } from 'node:net';
import { mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { afterEach, beforeEach, describe, expect, it } from 'vitest';

const ROOT = fileURLToPath(new URL('..', import.meta.url));
const CAVEMAN = join(ROOT, 'bin', 'caveman');

type Run = { code: number; stdout: string; stderr: string };

let runDir: string;

/** Runs the CLI against a throwaway run directory. Never rejects: the exit code
 *  is the assertion target. */
function run(args: string[], env: Record<string, string> = {}): Promise<Run> {
  return new Promise((resolve) => {
    execFile(
      CAVEMAN,
      args,
      { env: { ...process.env, CAVEMAN_RUN_DIR: runDir, ...env }, timeout: 40_000 },
      (error, stdout, stderr) => {
        const code = error === null ? 0 : typeof error.code === 'number' ? error.code : 1;
        resolve({ code, stdout, stderr });
      },
    );
  });
}

async function freePort(): Promise<number> {
  const server = createServer();
  await new Promise<void>((resolve) => server.listen(0, resolve));
  const { port } = server.address() as AddressInfo;
  await new Promise<void>((resolve) => server.close(() => resolve()));
  return port;
}

function startServerOn(port: number, body: string): Promise<Server> {
  const server = createServer((_request, response) => {
    response.writeHead(200, { 'content-type': 'application/json' });
    response.end(body);
  });
  return new Promise((resolve) => server.listen(port, () => resolve(server)));
}

beforeEach(async () => {
  runDir = await mkdtemp(join(tmpdir(), 'caveman-cli-'));
});

afterEach(async () => {
  await rm(runDir, { recursive: true, force: true });
});

describe('usage', () => {
  it('lists the clients it found on disk', async () => {
    const result = await run(['help']);
    expect(result.code).toBe(0);
    expect(result.stdout).toContain('claude');
  });

  it('exits 2 with no command, since nothing was done', async () => {
    const result = await run([]);
    expect(result.code).toBe(2);
  });

  it('names an unknown command', async () => {
    const result = await run(['nope']);
    expect(result.code).toBe(2);
    expect(result.stderr).toContain('nope');
  });
});

describe('port resolution', () => {
  it('prefers PORT from the environment', async () => {
    const result = await run(['status'], { PORT: '9393' });
    expect(result.stdout).toContain('9393');
  });

  it('reports the server’s own message for an out-of-range port', async () => {
    const result = await run(['status'], { PORT: '99999' });
    expect(result.code).toBe(1);
    expect(result.stderr).toContain('reading PORT failed');
    expect(result.stderr).toContain('99999');
  });

  it('reports the server’s own message for a non-numeric port', async () => {
    const result = await run(['status'], { PORT: 'abc' });
    expect(result.code).toBe(1);
    expect(result.stderr).toContain('reading PORT failed');
  });
});

describe('lifecycle', () => {
  it('starts, reports status, and stops', async () => {
    const port = String(await freePort());
    const started = await run(['up'], { PORT: port });
    expect(started.code).toBe(0);
    expect(started.stdout).toContain(port);
    expect(existsSync(join(runDir, 'caveman.pid'))).toBe(true);

    const status = await run(['status'], { PORT: port });
    expect(status.code).toBe(0);
    expect(status.stdout).toContain('running');

    const stopped = await run(['down'], { PORT: port });
    expect(stopped.code).toBe(0);
    expect(existsSync(join(runDir, 'caveman.pid'))).toBe(false);
  });

  it('reports an already-running server rather than starting a second', async () => {
    const port = String(await freePort());
    await run(['up'], { PORT: port });
    const pidFile = join(runDir, 'caveman.pid');
    const first = await readFile(pidFile, 'utf8');

    const again = await run(['up'], { PORT: port });
    expect(again.code).toBe(0);
    expect(again.stdout).toContain('already running');
    expect(await readFile(pidFile, 'utf8')).toBe(first);

    await run(['down'], { PORT: port });
  });

  it('exits 0 when stopping something that is not running', async () => {
    const result = await run(['down'], { PORT: String(await freePort()) });
    expect(result.code).toBe(0);
    expect(result.stdout).toContain('not running');
  });

  it('clears a stale pid file instead of reporting it as running', async () => {
    const port = String(await freePort());
    await writeFile(join(runDir, 'caveman.pid'), '999999');

    const status = await run(['status'], { PORT: port });
    expect(status.stdout).toContain('not running');

    const stopped = await run(['down'], { PORT: port });
    expect(stopped.code).toBe(0);
    expect(existsSync(join(runDir, 'caveman.pid'))).toBe(false);
  });
});

describe('a port held by another process', () => {
  let foreign: Server | null = null;

  afterEach(async () => {
    if (foreign !== null) {
      await new Promise<void>((resolve) => foreign!.close(() => resolve()));
      foreign = null;
    }
  });

  it('refuses to start when the answer is not caveman', async () => {
    const port = await freePort();
    foreign = await startServerOn(port, '{"status":"ok"}');

    const result = await run(['up'], { PORT: String(port) });
    expect(result.code).toBe(3);
    expect(result.stderr).toContain(String(port));
  });

  it('leaves that process alive', async () => {
    const port = await freePort();
    foreign = await startServerOn(port, '{"status":"ok"}');

    await run(['up'], { PORT: String(port) });
    expect(foreign.listening).toBe(true);
  });
});

describe('client dispatch', () => {
  let clientDir: string;

  beforeEach(async () => {
    clientDir = await mkdtemp(join(tmpdir(), 'caveman-clients-'));
  });

  afterEach(async () => {
    await rm(clientDir, { recursive: true, force: true });
  });

  it('passes the base url and every argument to the client', async () => {
    const port = String(await freePort());
    await writeFile(
      join(clientDir, 'echoer.sh'),
      'client_launch() { printf "url=%s args=%s\\n" "$CAVEMAN_BASE_URL" "$*"; }\n',
    );

    await run(['up'], { PORT: port });
    const result = await run(['echoer', '--resume', '-p', 'hi there'], {
      PORT: port,
      CAVEMAN_CLIENT_DIR: clientDir,
    });
    await run(['down'], { PORT: port });

    expect(result.stdout).toContain(`url=http://localhost:${port}`);
    expect(result.stdout).toContain('args=--resume -p hi there');
  });

  it('lists a client dropped into the directory, with no registry to edit', async () => {
    await writeFile(join(clientDir, 'codex.sh'), 'client_launch() { :; }\n');
    const result = await run(['help'], { CAVEMAN_CLIENT_DIR: clientDir });
    expect(result.stdout).toContain('codex');
  });
});

describe('compression level', () => {
  let clientDir: string;
  let port: string;

  /** Reports the level it would send and the arguments it was left with. */
  async function writeReporter(): Promise<void> {
    await writeFile(
      join(clientDir, 'show.sh'),
      'client_launch() { printf "level=%s args=%s\\n" "$CAVEMAN_LEVEL" "$*"; }\n',
    );
  }

  function show(args: string[]): Promise<Run> {
    return run(['show', ...args], { PORT: port, CAVEMAN_CLIENT_DIR: clientDir });
  }

  beforeEach(async () => {
    clientDir = await mkdtemp(join(tmpdir(), 'caveman-level-'));
    port = String(await freePort());
    await writeReporter();
  });

  afterEach(async () => {
    await run(['down'], { PORT: port });
    await rm(clientDir, { recursive: true, force: true });
  });

  it('rejects a level that is not one of the four', async () => {
    const result = await run(['up', '--level', 'bogus'], { PORT: port });
    expect(result.code).toBe(2);
    expect(result.stderr).toContain('bogus');
  });

  it('rejects the flag with no level after it', async () => {
    const result = await run(['up', '--level'], { PORT: port });
    expect(result.code).toBe(2);
  });

  it('compresses when no level was given, since asking for the CLI is the ask', async () => {
    await run(['up'], { PORT: port });
    expect((await show([])).stdout).toContain('level=caveman');
  });

  it('still allows off explicitly, for a baseline through the same proxy', async () => {
    await run(['up', '-l', 'off'], { PORT: port });
    expect((await show([])).stdout).toContain('level=off');
  });

  it('inherits the level given to up', async () => {
    await run(['up', '-l', 'moderate'], { PORT: port });
    expect((await show([])).stdout).toContain('level=moderate');
  });

  it('reports the stored level in status', async () => {
    await run(['up', '-l', 'light'], { PORT: port });
    const result = await run(['status'], { PORT: port });
    expect(result.stdout).toContain('level: light');
  });

  it('lets a client override the stored level for that launch only', async () => {
    await run(['up', '-l', 'moderate'], { PORT: port });
    expect((await show(['-l', 'caveman'])).stdout).toContain('level=caveman');
    expect((await show([])).stdout).toContain('level=moderate');
  });

  it('accepts --level=value as well as a separate argument', async () => {
    await run(['up', '--level=caveman'], { PORT: port });
    expect((await show([])).stdout).toContain('level=caveman');
  });

  it('keeps the level flag out of the arguments the client receives', async () => {
    await run(['up'], { PORT: port });
    const result = await show(['-l', 'light', '--resume', '-p', 'hi there']);
    expect(result.stdout).toContain('args=--resume -p hi there');
  });

  it('leaves a client’s own -l alone after --', async () => {
    await run(['up', '-l', 'light'], { PORT: port });
    const result = await show(['--', '-l', 'something']);
    expect(result.stdout).toContain('level=light');
    expect(result.stdout).toContain('args=-l something');
  });
});

File: internal/cli/cli_test.go
package cli_test

import (
	"os"
	"path/filepath"
	"testing"
)

func TestUsageListsTheClientsItFound(t *testing.T) {
	result := newHarness(t).run("help")
	requireCode(t, result, 0)
	requireContains(t, result.stdout, "claude", "usage")
}

func TestNoCommandExitsTwo(t *testing.T) {
	requireCode(t, newHarness(t).run(), 2)
}

func TestUnknownCommandIsNamed(t *testing.T) {
	result := newHarness(t).run("nope")
	requireCode(t, result, 2)
	requireContains(t, result.stderr, "nope", "unknown command")
}

func TestPortPrefersTheEnvironment(t *testing.T) {
	result := newHarness(t).with("PORT", "9393").run("status")
	requireContains(t, result.stdout, "9393", "status")
}

func TestOutOfRangePortReportsTheServersMessage(t *testing.T) {
	result := newHarness(t).with("PORT", "99999").run("status")
	requireCode(t, result, 1)
	requireContains(t, result.stderr, "reading PORT failed", "port failure")
	requireContains(t, result.stderr, "99999", "port failure")
}

func TestNonNumericPortReportsTheServersMessage(t *testing.T) {
	result := newHarness(t).with("PORT", "abc").run("status")
	requireCode(t, result, 1)
	requireContains(t, result.stderr, "reading PORT failed", "port failure")
}

func TestStartsReportsStatusAndStops(t *testing.T) {
	harness := newHarness(t)
	port := freePort(t)
	harness.with("PORT", port)
	harness.stopOnCleanup(port)

	started := harness.run("up")
	requireCode(t, started, 0)
	requireContains(t, started.stdout, port, "up")
	if !exists(harness.path("caveman.pid")) {
		t.Fatal("expected a pid file after up")
	}

	status := harness.run("status")
	requireCode(t, status, 0)
	requireContains(t, status.stdout, "running", "status")

	stopped := harness.run("down")
	requireCode(t, stopped, 0)
	if exists(harness.path("caveman.pid")) {
		t.Fatal("expected the pid file gone after down")
	}
}

func TestReportsAnAlreadyRunningServer(t *testing.T) {
	harness := newHarness(t)
	port := freePort(t)
	harness.with("PORT", port)
	harness.stopOnCleanup(port)

	harness.run("up")
	first, err := os.ReadFile(harness.path("caveman.pid"))
	if err != nil {
		t.Fatal(err)
	}

	again := harness.run("up")
	requireCode(t, again, 0)
	requireContains(t, again.stdout, "already running", "second up")

	second, err := os.ReadFile(harness.path("caveman.pid"))
	if err != nil {
		t.Fatal(err)
	}
	if string(second) != string(first) {
		t.Fatalf("pid changed: %q then %q", first, second)
	}
}

func TestStoppingWhatIsNotRunningExitsZero(t *testing.T) {
	result := newHarness(t).with("PORT", freePort(t)).run("down")
	requireCode(t, result, 0)
	requireContains(t, result.stdout, "not running", "down")
}

func TestClearsAStalePIDFile(t *testing.T) {
	harness := newHarness(t).with("PORT", freePort(t))
	if err := os.WriteFile(harness.path("caveman.pid"), []byte("999999"), 0o644); err != nil {
		t.Fatal(err)
	}

	status := harness.run("status")
	requireContains(t, status.stdout, "not running", "status")

	stopped := harness.run("down")
	requireCode(t, stopped, 0)
	if exists(harness.path("caveman.pid")) {
		t.Fatal("expected the stale pid file removed")
	}
}

func TestRefusesToStartWhenTheAnswerIsNotCaveman(t *testing.T) {
	port := freePort(t)
	startForeign(t, port, `{"status":"ok"}`)

	result := newHarness(t).with("PORT", port).run("up")
	requireCode(t, result, 3)
	requireContains(t, result.stderr, port, "port taken")
}

func TestLeavesTheForeignProcessAlive(t *testing.T) {
	port := freePort(t)
	startForeign(t, port, `{"status":"ok"}`)

	newHarness(t).with("PORT", port).run("up")

	// Its own body is the proof it was neither closed nor replaced.
	body := getBody(t, "http://127.0.0.1:"+port+"/health")
	if body != `{"status":"ok"}` {
		t.Fatalf("foreign server changed: %q", body)
	}
}

// clientDir writes a drop-in client and points the CLI at its directory.
func clientDir(t *testing.T, name, script string) string {
	t.Helper()
	dir := t.TempDir()
	path := filepath.Join(dir, name)
	if err := os.WriteFile(path, []byte(script), 0o755); err != nil {
		t.Fatal(err)
	}
	return dir
}

const echoerScript = `#!/bin/sh
printf "url=%s args=%s\n" "$CAVEMAN_BASE_URL" "$*"
`

const reporterScript = `#!/bin/sh
printf "level=%s args=%s header=%s\n" "$CAVEMAN_LEVEL" "$*" "$CAVEMAN_COMPRESS_HEADER"
`

func TestPassesTheBaseURLAndEveryArgumentToTheClient(t *testing.T) {
	harness := newHarness(t)
	port := freePort(t)
	harness.with("PORT", port)
	harness.with("CAVEMAN_CLIENT_DIR", clientDir(t, "echoer", echoerScript))
	harness.stopOnCleanup(port)

	harness.run("up")
	result := harness.run("echoer", "--resume", "-p", "hi there")

	requireContains(t, result.stdout, "url=http://localhost:"+port, "client env")
	requireContains(t, result.stdout, "args=--resume -p hi there", "client args")
}

func TestListsAClientDroppedIntoTheDirectory(t *testing.T) {
	harness := newHarness(t).with("CAVEMAN_CLIENT_DIR", clientDir(t, "codex", echoerScript))
	result := harness.run("help")
	requireContains(t, result.stdout, "codex", "usage")
}

// levelHarness is a run directory with a reporting client installed.
func levelHarness(t *testing.T) (*harness, string) {
	t.Helper()
	harness := newHarness(t)
	port := freePort(t)
	harness.with("PORT", port)
	harness.with("CAVEMAN_CLIENT_DIR", clientDir(t, "show", reporterScript))
	harness.stopOnCleanup(port)
	return harness, port
}

func TestRejectsALevelThatIsNotOneOfTheFour(t *testing.T) {
	harness, _ := levelHarness(t)
	result := harness.run("up", "--level", "bogus")
	requireCode(t, result, 2)
	requireContains(t, result.stderr, "bogus", "level rejection")
}

func TestRejectsTheFlagWithNoLevelAfterIt(t *testing.T) {
	harness, _ := levelHarness(t)
	requireCode(t, harness.run("up", "--level"), 2)
}

func TestCompressesWhenNoLevelWasGiven(t *testing.T) {
	harness, _ := levelHarness(t)
	harness.run("up")
	requireContains(t, harness.run("show").stdout, "level=caveman", "default level")
}

func TestStillAllowsOffExplicitly(t *testing.T) {
	harness, _ := levelHarness(t)
	harness.run("up", "-l", "off")
	requireContains(t, harness.run("show").stdout, "level=off", "off level")
}

func TestInheritsTheLevelGivenToUp(t *testing.T) {
	harness, _ := levelHarness(t)
	harness.run("up", "-l", "moderate")
	requireContains(t, harness.run("show").stdout, "level=moderate", "inherited level")
}

func TestReportsTheStoredLevelInStatus(t *testing.T) {
	harness, _ := levelHarness(t)
	harness.run("up", "-l", "light")
	requireContains(t, harness.run("status").stdout, "level: light", "status level")
}

func TestClientOverridesTheStoredLevelForThatLaunchOnly(t *testing.T) {
	harness, _ := levelHarness(t)
	harness.run("up", "-l", "moderate")
	requireContains(t, harness.run("show", "-l", "caveman").stdout, "level=caveman", "override")
	requireContains(t, harness.run("show").stdout, "level=moderate", "after override")
}

func TestAcceptsLevelEqualsValue(t *testing.T) {
	harness, _ := levelHarness(t)
	harness.run("up", "--level=caveman")
	requireContains(t, harness.run("show").stdout, "level=caveman", "--level=value")
}

func TestKeepsTheLevelFlagOutOfTheClientArguments(t *testing.T) {
	harness, _ := levelHarness(t)
	harness.run("up")
	result := harness.run("show", "-l", "light", "--resume", "-p", "hi there")
	requireContains(t, result.stdout, "args=--resume -p hi there", "stripped flag")
}

func TestLeavesAClientsOwnDashLAloneAfterDoubleDash(t *testing.T) {
	harness, _ := levelHarness(t)
	harness.run("up", "-l", "light")
	result := harness.run("show", "--", "-l", "something")
	requireContains(t, result.stdout, "level=light", "level after --")
	requireContains(t, result.stdout, "args=-l something", "args after --")
}

// Version is what a bug report identifies a build by. It is stamped with
// -ldflags at release; a build made any other way says "dev" rather than
// reporting nothing at all.
func TestVersionReportsTheBuild(t *testing.T) {
	for _, command := range []string{"version", "--version", "-v"} {
		result := newHarness(t).run(command)
		requireCode(t, result, 0)
		requireContains(t, result.stdout, "caveman", "dev")
	}
}
